### ISSUE TITLE
Access violation when returing object with specific layout on MSVC

### ISSUE THREAD

**Issue by Lyulf** [boostorg/cobalt#172]
Code:
```c++
#include <cstdint>
#include <cstdio>
#include <string>

#include <boost/cobalt.hpp>

namespace cobalt = boost::cobalt;

struct ExampleStruct
{
  std::string str;
  std::uint8_t padding[32];
};

cobalt::task<ExampleStruct> getResult()
{
  // error on this return
  co_return {};
}

cobalt::main co_main(int /*argc*/, char** /*argv*/)
{
  co_await getResult();
  // Never displayed
  std::puts("Done");
  co_return 0;
}

```

Built on MSVC 17.9.5 (cl version 19.39.33523 for x64) with following command:
```
cl /Femain.exe main.cpp /nologo /external:IC:\Users\chodorowskip\repo\deps\vc143\boost-1.85.0\include\boost-1_85 /std:c++20 /MD /Z7 /O2 /EHsc /D_WIN32_WINNT=0x0601 /W4 /external:W0 /link /PROFILE /LIBPATH:C:\Users\chodorowskip\repo\deps\vc143\boost-1.85.0\lib libboost_cobalt-vc143-mt-x64-1_85.lib
```

Error
```
Exception thrown at 0x00007FF728E265E1 in main.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
```

Error doesn't seem to happen with debug configuration.

**Comment by Pele44** [comment]
Is there a chance that the error is actually in cobalt? Or maybe it's better to report the case to Microsoft? The problem occurs in our project and unfortunately the current workaround is to stop using cobalt, which we would really like not to do, because it is supposed to be a very helpful library, making it easier to solve many of the problems we face.

**Comment by klemens-morgenstern** [comment]
I had plenty of problems with MSVC and last time I checked they had issues with improper inlining (for which I added a workaround). I however don't like to say "it's their fault" as it might be mine. I didn't get around to checking the code yet - which is why I didn't respond, sorry.

**Comment by Pele44** [comment]
Thanks for the answer!

**Comment by fgoujeon** [comment]
It this related to this [MSVC bug](https://developercommunity.visualstudio.com/t/coroutine-Access-violation-reading-locat/1383447?q=+Cannot+read+properties+of+undefined+%28reading+%27coreFields%27%29)?

**Comment by klemens-morgenstern** [comment]
It might be. I added commit increasing the align [here](https://github.com/boostorg/cobalt/tree/max_align). 

I don't know when I get my windows builds working again, so if you get around to test it let me know.

**Comment by Lyulf** [comment]
I couldn't figure out how to build cobalt without bootstrap and b2, so I just applied the changes from the most recent commit on the branch you provided and they don't seem to work. Still got access violation.

**Comment by Pele44** [comment]
Is there a chance that the mentioned bug/workaround for the MSVC bug will be fixed in the near future?

**Comment by Mishura4** [comment]
Spent a lot of time converting a lot of our async callbacks to coroutines, only to find out this happens in release.

Any way this can be looked at?

**Comment by klemens-morgenstern** [comment]
Which boost version did you build with?

**Comment by konradglas** [comment]
@klemens-morgenstern I tested [this branch ](https://github.com/boostorg/cobalt/tree/max_align)
It still crashes for me at the exact same place. It definitely is the msvc compiler bug that assumes 16 byte aligment (according to the post) . What works for me is to over-align the promise_type itself like so:
```
struct alignas(32) task_promise : promise_memory_resource_base,
                                  promise_cancellation_base<asio::cancellation_slot, asio::enable_total_cancellation>,
                                  promise_throw_if_cancelled_base,
                                  enable_awaitables<task_promise<Return>>,
                                  enable_await_allocator<task_promise<Return>>,
                                  enable_await_executor<task_promise<Return>>,
                                  enable_await_deferred,
                                  task_promise_result<Return>
{
```
Note that `alignas(16)` doesn't seem to cut it.

**Comment by klemens-morgenstern** [comment]
Thank you for that. I hopefully find the time to look into that, here's 

If seems this has been fixed in a recent MSVC release:
https://developercommunity.visualstudio.com/t/C20-coroutine-variables-are-not-proper/10049371

I am not sure how to work around this - probably a version check (if the fix works) and then do a 32 align.

I just wish they fixed this one:
https://developercommunity.visualstudio.com/t/Unable-to-destroy-C20-coroutine-in-fin/10657377

**Comment by konradglas** [comment]
yeah ... weird and sad that they are not fixing this. I have tried with the latest msvc (17.14) and it still crashes. So no :( they didn't fix that particular part. FWIW this would be a fix that compiles for msvc, latest apple-clang and GCC (Ubuntu)
```
#ifdef _MSC_VER
#define OVERALIGN alignas(32)
#else
#define OVERALIGN
#endif

struct async_initiate_spawn;

template <typename Return>
struct OVERALIGN task_promise : promise_memory_resource_base,
```
regarding this bug [https://developercommunity.visualstudio.com/t/Unable-to-destroy-C20-coroutine-in-fin/10657377](https://developercommunity.visualstudio.com/t/Unable-to-destroy-C20-coroutine-in-fin/10657377)
before I run into an msvc coro quirk again. Does this have any implications when running cobalt? Should I avoid something in particular?

**Comment by klemens-morgenstern** [comment]
I'll need to add a `BOOST_COBALT_MSVC_ALIGN_PROMISE` I think, and then users can override it. the 32 is for the 64 bit compiler? I reckon 32 byte might work for the 32 bit then.

The other bug is worked around, the self delete does a post to the event loop. I.e. this is slower than on clang/gcc, but works. The general issues msvc has is bad inlining. I.e. it can inline function calls into the coroutine frame which can lead to horrible bugs. I am using a bunch of `noinline` in many places in the code, but I might have missed something - it's hard to test for compiler errors.

**Comment by konradglas** [comment]
Yes I was fearing that as well. 
I am also very strict for that reason regarding inlining on msvc.

This is for the 64 bit compiler. I cannot check but I thought this bug only occurs for 64 bit.
